Add `--retry-until-up`, `--region`, `--zone`, and `--idle-minutes-to-autostop` for interactive nodes (v2)

ewzeng opened the pull request:
This pull request builds on https://github.com/skypilot-org/skypilot/pull/1207 by setting `user_requested_resources` appropriately and handling the autostop edge case.

Tested:

`sky {cpunode, gpunode}` on GCP:

- [x] With and without -`-idle-minutes-to-autostop`
- [x] With varying regions and zones (`--region`, `--zone`)

I used print statements to verify `_launch_with_confirm` is called with the correct `retry_until_up` argument. 
I tested the autostop edge case by running:

- [x] `sky {cpunode, gpunode}` followed by `sky {cpunode, gpunode} -i 1`

I tested that `user_requested_resources` was set correctly by running:

- [x] `sky {cpunode, gpunode}` followed by `sky {cpunode, gpunode}` with  `--region` and `--zone` flags

@concretevitamin @Michaelvll

Michaelvll commented on sky/cli.py:
```diff
@@ -208,15 +208,53 @@ def _interactive_node_cli_command(cli_func):
                               default=False,
                               required=False,
                               help='Skip confirmation prompt.')
+    idle_autostop = click.option('--idle-minutes-to-autostop',
+                                 '-i',
+                                 default=None,
+                                 type=int,
+                                 required=False,
+                                 help=('Automatically stop the cluster after '
+                                       'this many minutes of idleness, i.e. '
+                                       'no running or pending jobs in the '
+                                       'cluster\'s job queue. Idleness starts '
+                                       'counting after setup/file_mounts are '
+                                       'done; the clock gets reset whenever '
+                                       'there are running/pending jobs in the '
+                                       'job queue. If not set, the cluster '
+                                       'will not be auto-stopped.'))
+    retry_until_up = click.option('--retry-until-up',
+                                  '-r',
+                                  is_flag=True,
+                                  default=False,
+                                  required=False,
+                                  help=('Whether to retry provisioning '
+                                        'infinitely until the cluster is up '
+                                        'if we fail to launch the cluster on '
+                                        'any possible region/cloud due to '
+                                        'unavailability errors.'))
+    region_option = click.option('--region',
+                                 default=None,
+                                 type=str,
+                                 required=False,
+                                 help='The region to use.')
+    zone_option = click.option('--zone',
+                               default=None,
+                               type=str,
+                               required=False,
+                               help='The zone to use.')
```
Instead of duplicating the options here, can we directly use the `_TASK_OPTIONS` or split the `_TASK_OPTIONS` to `_TASK_OPTIONS` and `_CLUSTER_OPTIONS`, so that both `launch` and `{cpu,gpu,tpu}node` can share the similar options (including the cloud, instance_type, ... options above)?

ewzeng replied:
One drawback is that the descriptions in `_TASK_OPTIONS` refer to config in the YAML, but `sky {cpu,gpu,tpu} node` don't use the YAML.

Michaelvll replied:
Good point! Do you think it is possible to move the construction of the `_TASK_OPTIONS` to a function?
```
def _get_task_options(has_yaml: bool) -> List[click.Option]
```

ewzeng replied:
Sure, let me try this.

ewzeng replied:
@Michaelvll I have a question. Is `If specified, overrides the "resources.cloud" config.` referring to the yaml config? I thought it was, but there are other help messages explicitly refer to the yaml config (like `Overrides the "num_nodes" config in the YAML if both are supplied.`)

Michaelvll replied:
Yes, they should all refer to the YAML config. Great catch! We should make them consistent, e.g. `If specified, overrides the "resources.cloud" config in the YAML.` and ``If specified, overrides the "num_nodes" config in the YAML.`

Michaelvll commented on sky/cli.py:
```diff
@@ -657,6 +695,13 @@ def _launch_with_confirm(
 
+    if node_type is not None and idle_minutes_to_autostop is not None \
+            and maybe_status == global_user_state.ClusterStatus.UP:
+        # We do not sky.launch if interactive node is already up, so we need
+        # to update idle timeout here.
+        handle = global_user_state.get_handle_from_cluster_name(cluster)
+        backend.set_autostop(handle, idle_minutes_to_autostop)
```
Great catch! How about we use the programmatic API `core.autostop` instead?

Also, should we make this an `elif idle_minutes_to_autostop is not None` after L718?

Michaelvll commented on sky/cli.py:
```diff
@@ -208,15 +208,53 @@ def _interactive_node_cli_command(cli_func):
                               default=False,
                               required=False,
                               help='Skip confirmation prompt.')
+    idle_autostop = click.option('--idle-minutes-to-autostop',
```
`--down` option is also needed to decide whether this is `autostop` or `autodown`

Michaelvll reviewed (commented):
Thank you for the fixes @ewzeng! Those options would be very useful for the interactive nodes. Left several comments.

ewzeng commented:
Thanks for the comments @Michaelvll! I will work on this.

ewzeng commented:
@Michaelvll I've pushed a commit that refactors the click options! I haven't tested it extensively, but it would be great to get some feedback! Is this what you had in mind?

Michaelvll commented:
> @Michaelvll I've pushed a commit that refactors the click options! I haven't tested it extensively, but it would be great to get some feedback! Is this what you had in mind?

Thank you for the quick fix @ewzeng! After reading the new changes, I found the readability of it is a bit harder than before. Sorry for bringing this up, but I found it may be better to keep the original version.

ewzeng commented:
No worries, I think I agree haha. I will revert.

ewzeng commented:
I have implemented `--down` and switched to `core.autostop`, as you suggested.

Tested:
- [x] `sky {cpunode, gpunode}` followed by `sky {cpunode, gpunode} -i 10`
- [x] `sky {cpunode, gpunode}` followed by `sky {cpunode, gpunode} --down`
- [x] `sky {cpunode, gpunode}` followed by `sky {cpunode, gpunode} -i 10 --down`
- [x] `sky {cpunode, gpunode} --down`

Michaelvll reviewed (approved):
Thank you for fixing this @ewzeng! LGTM.

This can be another PR: We have a problem running `sky gpunode` to log into the existing cluster launched by `sky gpunode --gpus V100 --cloud azure`. Do you want to take a look at it? https://github.com/skypilot-org/skypilot/issues/1331

ewzeng commented:
Thanks! @Michaelvll! You may need to merge for me (I don't have write access).

I will take a look at #1331

concretevitamin commented:
Thanks @ewzeng!
